src/mappings/mappingUniquesHandlers.ts

```diff
@@ -1,28 +1,145 @@
   const to = event.event.data[3];
   const collectionId = event.event.data[0];
   const itemId = event.event.data[1];
+
+  if (!from || !to || !collectionId || !itemId) {
+    logger.debug(
+      "Some arguments is null",
+      JSON.stringify(event.toHuman())
+    );
+    return;
+  }
+
   const uniqueTransfer = new UniquesTransfer(
     `${event.block.block.header.number.toNumber()}-${event.idx}`
   );
+
   uniqueTransfer.block = event.block.block.header.number.toNumber();
   uniqueTransfer.from = from.toString();
   uniqueTransfer.to = to.toString();
-  uniqueTransfer.collectionId = collectionId.toString();
-  uniqueTransfer.itemId = itemId.toString();
   if (event.extrinsic) {
+    event.extrinsic.events.forEach((eventItem) => {
+      if (eventItem.event.method === "Withdraw") {
+        const fee = eventItem.event.data[1];
+        uniqueTransfer.fee = (fee as Balance).toBigInt();
+      }
+    });
     uniqueTransfer.txHash = event.extrinsic.extrinsic.hash.toString();
     uniqueTransfer.timestamp = new Date(
-    event.extrinsic.block.timestamp
-  ).getTime();
+      event.extrinsic.block.timestamp
+    ).getTime();
   }
 
+  const collection = await ensureCollection(collectionId.toString());
+  const item = await ensureItem(collectionId.toString(), itemId.toString());
+  item.owner = to.toString();
+  uniqueTransfer.itemId = item.id;
+  uniqueTransfer.collectionId = collection.id;
+
+  await collection.save();
+  await item.save()
   return uniqueTransfer.save();
 }
+
+export const handleUniquesMetadataSetEvent = async (
+  event: SubstrateEvent
+) => {
+  logger.debug("uniqueMetadataSetEvent added: " + JSON.stringify(event.toHuman()));
+  const collectionId = event.event.data[0];
+  const itemId = event.event.data[1];
+  const data = event.event.data[2];
+  // const isFrozen = event.event.data[3];
+
+  const collection = await ensureCollection(collectionId.toString());
+  const item = await ensureItem(collectionId.toString(), itemId.toString());
+  item.metadataCid = data.toHuman().toString();
+  item.collectionId = collection.id;
+  await collection.save()
+  return item.save();
+}
+
+export const handleUniquesCollectionMetadataSetEvent = async (
+  event: SubstrateEvent
+) => {
+  logger.debug(
+    "uniqueCollectionMetadataSetEvent added: " + JSON.stringify(event.toHuman())
+  );
+  const collectionId = event.event.data[0];
+  const data = event.event.data[1];
+  // const isFrozen = event.event.data[2];
+
+  const collection = await ensureCollection(collectionId.toString());
+  collection.metadataCid = data.toHuman().toString();
+
+  return collection.save();
+}
+
+export const handleUniquesDestroyedEvent = async (
+  event: SubstrateEvent
+) => {
+  logger.debug(
+    "handleUniquesDestroyedEvent added: " + JSON.stringify(event.toHuman())
+  );
+  const collectionId = event.event.data[0];
+
+  const collection = await ensureCollection(collectionId.toString());
+  collection.isDestroyed = true;
+  return collection.save()
```

Comment:
Could it be a bette choice to:
```suggestion
  return Collection.remove(collectionId.toString());
```
?

Reply:
Not really, since the items and collections have relation with transaction history so on delete it could cause an rule violation
We need the transaction history to be persistent

Reply:
It would then tell me that we will have issues when a collection is destroyed and then the same id is used to create another collection. We may then link the transaction history to the wrong collection. Isn't it?
I'm thinking of an approach to avoid this situation. One way of doing this could be that we record the block number that a collection was created in as part of the collection's unique id. We need to make a new unique id for collections which is a made like `S{blockNumber}-${collectionId}`. Whenever there is an event changing a collection, we should only change the relevant collection. And to find the relevant collection, the collection id by itself is not enough.

Reply:
True Alex, I had that on the notes but did not fix it... will prepare a hot fix for tomorrow